eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/ChannelTlv.scala

```diff
@@ -29,23 +29,34 @@ sealed trait OpenChannelTlv extends Tlv
 
 sealed trait AcceptChannelTlv extends Tlv
 
+sealed trait OpenChannel2Tlv extends Tlv
+
+sealed trait AcceptChannel2Tlv extends Tlv
+
 object ChannelTlv {
 
   /** Commitment to where the funds will go in case of a mutual close, which remote node will enforce in case we're compromised. */
-  case class UpfrontShutdownScriptTlv(script: ByteVector) extends OpenChannelTlv with AcceptChannelTlv {
+  case class UpfrontShutdownScriptTlv(script: ByteVector) extends OpenChannelTlv with AcceptChannelTlv with OpenChannel2Tlv with AcceptChannel2Tlv {
     val isEmpty: Boolean = script.isEmpty
   }
 
   val upfrontShutdownScriptCodec: Codec[UpfrontShutdownScriptTlv] = variableSizeBytesLong(varintoverflow, bytes).as[UpfrontShutdownScriptTlv]
 
   /** A channel type is a set of even feature bits that represent persistent features which affect channel operations. */
-  case class ChannelTypeTlv(channelType: ChannelType) extends OpenChannelTlv with AcceptChannelTlv
+  case class ChannelTypeTlv(channelType: ChannelType) extends OpenChannelTlv with AcceptChannelTlv with OpenChannel2Tlv with AcceptChannel2Tlv
 
   val channelTypeCodec: Codec[ChannelTypeTlv] = variableSizeBytesLong(varintoverflow, bytes).xmap(
     b => ChannelTypeTlv(ChannelTypes.fromFeatures(Features(b).initFeatures())),
     tlv => Features(tlv.channelType.features.map(f => f -> FeatureSupport.Mandatory).toMap).toByteVector
   )
 
+  case class ChannelFlagsTlv(announceChannel: Boolean) extends OpenChannel2Tlv with AcceptChannel2Tlv
+
+  val channelFlagsCodec: Codec[ChannelFlagsTlv] = variableSizeBytesLong(varintoverflow, bytes).xmap(
+    b => ChannelFlagsTlv(b.lastOption.exists(v => (v % 2) == 1)),
+    tlv => if (tlv.announceChannel) ByteVector(1.toByte) else ByteVector(0.toByte)
+  )
```

Comment:
> Btw, what's the rationale for making this field a tlv?

It's more future-proof if we want to introduce additional channel flags in the future.
Now that we're re-working the messages, it's a good opportunity to do it.
Even if we never use it, it doesn't cost us anything to have this flexibility.

> Shouldn't you use the ChannelFlags class and codec?

It will deviate once we add new channel flags (which will probably only apply to dual funded channels).
I can rename that `DualFundedChannelFlags` for clarity, how does that sound?

Reply:
I prefer `DualFundedChannelFlags` also for clarify.

Reply:
Done in https://github.com/ACINQ/eclair/pull/2231/commits/dabcbbba4ca4d2169f16105e6755aae2619162c0

Reply:
> It's more future-proof if we want to introduce additional channel flags in the future.

We could use a variable length field? I guess to me tlvs are for additional features, and channel flags is a core part of the `open_message`. Another way to look at it: why not make all fields of `open_channel2` tlvs?

Reply:
> Another way to look at it: why not make all fields of open_channel2 tlvs?

I proposed this on the spec PR, but it was rejected. The way we should create new messages is to put every field that is required all the time in the core of the message and the rest in tlvs. You're right that `channel_flags` make sense in the core fields, there's no good reason to make it a tlv. I'll move it to the core of the message with a `length` prefix and we'll see what niftynei says.

Reply:
Done in https://github.com/ACINQ/eclair/pull/2231/commits/a7c2d3686c047502634edebfcfc3165ab404ac4a and updated in https://github.com/niftynei/lightning-rfc/pull/3